### ISSUE TITLE
Run Doctest inside modules fails inside trait implementations for primitive types and types with non-alphanumerical characters

### ISSUE THREAD

**Issue by kangalioo** [rust-analyzer/rust-analyzer#8248]
When you have a trait implementation for a type inside a module (i.e. not crate root), for some types the Run Doctest lens at the top of doctests for trait methods fails to run any test.

It seems that affected types are primitive types (str, u32, char...) and types with non-alphanumerical characters in them (`&T`, `(T, T)`, `[T; N]`...).

Curiously enough, the issue only seems to encounter within modules; when the trait implementation is moved to the crate root, Run Doctest does its job just fine regardless of the type used (as far as I could tell).

Example:

```rust
mod hi {
    pub trait Trait {
        fn foo();
    }
    
    impl Trait for char {
        /// ```rust
        /// assert_eq!(1 + 1, 2);
        /// ```
        fn foo() {}
    }
}
```

![Screenshot_20210330_015905](https://user-images.githubusercontent.com/21220820/112914154-88b9c880-90fb-11eb-898b-365dbdf460ed.png)

The issue seems to be that the type name is, for some reason, dropped from the path passed to `--package`.

**Comment by Veykril** [comment]
Yep, we are only handling adt's here currently https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/runnables.rs#L302

**Comment by darrow-olykos** [comment]
@Veykril 

Just ran into this bug as well, would you be able to leave any sort of mental breadcrumbs for me if I wanted to take a stab at resolving? I see this bug appeared due to a refactoring of this function ( https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/runnables.rs#L355 )

What do I need to understand about this `adt` to start to try to fix?

**Comment by Veykril** [comment]
ADT = Abstract-Data Type, meaning struct, enum and union types.

What we need at that location is the name of the type so that we can emit it for the runnable command, hence why we check if the type is an ADT, since then we can fetch the name of it.
This fails for primitive types since those are not ADTs, so we need an extra if branch here that checks for whether the type is a primitive type. I think there should be a similar function defined like the `as_adt` for these. Then we just need to fetch the name of the primitive there.

**Comment by darrow-olykos** [comment]
@Veykril Hm, I think I see.
I could add a local helper `fn` like this:

```rust
fn get_name(type: Type) -> Option<Name> {
    if let Some(adt) = ty.as_adt() {
        adt.name(db);
    } else if let Some(builtin) = ty.as_builtin() {
        builtin.name();
    }
    else {
        None
    }
}
```

And I could update this function to look like:

```
    let name = ty.get_name()?;
```

Do you think this helper function belongs better on the `Type` within `hir` crate instead?

**Comment by darrow-olykos** [comment]
Mkay, I could do something like
```rust
                let name = match ty.as_adt() {
                    Some(adt) => adt().name(db),
                    None => match ty.as_builtin() {
                        Some(builtin) => builtin.name(db),
                        None => None,
                    }
                };
 ```
 
 Question, why do I see `adt.name(db)` taking in `db` if the `HasName` trait's `name` method only has `&self`?

**Comment by darrow-olykos** [comment]
Are there any docs in this repo that explain how to open a vscode project that uses a local rust-analyzer repo rather than an officially installed rust-analyzer extension? @Veykril

**Comment by Veykril** [comment]
running `cargo xtask install --server` in the rust-analyzer workspace will build and install the rust-analyzer server in your cargo binaries, which if they are in your `PATH` can be then executed in VSCode by setting the `rust-analyzer.server.path` configuration to `rust-analyzer`.

**Comment by darrow-olykos** [comment]
Ok so if do the following:
1) run `cargo xtask install --server` in my rust-analyzer workspace
2) add "analyzer.server.path": "rust-analyzer"
3) ensure cargo binaries are in PATH

Then will any workspace I open up with vscode be using my local rust-analyzer instead of official extension?
Do I need to use "Run Extension (Debug build)" with the above 1-3?

**Comment by Veykril** [comment]
Just with your 3 steps running a new VSCode instance should run the self-build server. Running the debug steps shouldnt be necessary. Though using one of the debug server tasks should also work(without the need of the steps I described before, that was just my workflow for these).

**Comment by Veykril** [comment]
Otherwise the only relevant doc I can point you to is https://github.com/rust-analyzer/rust-analyzer/tree/master/docs/dev#launching-rust-analyzer

**Comment by darrow-olykos** [comment]
@Veykril Is there any way to _know_ if vscode is in fact using local rusty-analyzer? For example, if I add `println!("visit_file_defs(_) ... !!!");` within the `visit_file_defs` callback FnMut within `runnables.rs`, where should I even see that print?

**Comment by Veykril** [comment]
Note sure if we show println, you want to use `tracing` to log instead and then set the `RA_LOG` env var to whatever tracing log level you want so `info` or `warn` for example(or for more finegrained logging `path::to::module=info`). The output then appears in `Rust Analyzer Langauge Server`(not Trace).

**Comment by darrow-olykos** [comment]
Hm... I have done the following:
1. set `RA_LOG=info` env var
2. Added the following to my settings.json in vscode: `"rust-analyzer.trace.server": "verbose"`
3. Added `tracing::info!("visit_file_defs {:?}", &file_id);` to `pub fn visit_file_defs`
4. Re-ran `cargo xtask install --server`
5. Opened new vscode workspace
6. Looked in dropdown in OUTPUT tab in my new window: I see only these two `Rust Analyzer` related options: `Rust Analyzer Client` or `Rust Analyzer Server Trace`. The `client` selection shows same output as before (just displays some configuration), and the `Rust Analyzer Server Trace` option is now outputting a lot of information, however NOT anything related to the `tracing::info` statement that I added.

RE: Your last message:

- _"`in Rust Analyzer Language Server` (not Trace)"_ -- There is no option for this, there is only `Rust Analyzer Server Trace` or `Rust Analyzer Client`.
- _"You can check what r-a server you are running by executing the Rust-Analyzer: Run command as well" -- Doing this seems to have no affect (does nothing)

Any thoughts?

**Comment by darrow-olykos** [comment]
@Veykril Your assumption is correct

RE: RA_LOG, good news is, you were right, adding that caused another selectable option under OUTPUT tab for `Rust Analyze Language Server`

The weird news is, I cannot see my tracing output still. Even moved it to `crates/rust-analyzer/src/bin/main.rs`. 
It seems I can trigger an error with `panic!`.. sort of. When I add a `panic!` instead of `tracing::info!` and then open another vscode workspace, I end up seeing this error:

```
ERROR [3/2/2022, 9:14:38 PM]: Bootstrap error [Error: Failed to execute rust-analyzer --version. `config.server.path` or `config.serverPath` has been set explicitly.            Consider removing this config or making a valid server binary available at that path.
```

However that's in `Rust Analyzer Client`, and I don't see my actual `panic!` message, and I also don't see the dropdown options for `Rust Analyzer Server` and `Rust Analyzer Server Traces` under the OUTPUT tab.

So... I'm... _kind of_ convinced that I'm using the local rust-analyzer repo for this extension but... I still need to figure out how to see my manual tracing info though.

**Comment by MariaSolOs** [comment]
So I just ran into this issue with types implemented in the standard library.

Here are my repro steps:
1. Inside `src/main.rs` include the following:
```
use std::sync::Mutex;

fn main() {
    let m = Mutex::new(0);
}
```
2. Place your cursor (`|`) in `Mutex::ne|w` and trigger "Go to definition" (A.K.A. F12).
3. VS Code will display the code lens option to "Run Doctest" in the library's definition:
<img width="773" alt="image" src="https://user-images.githubusercontent.com/62502207/203462899-801fbe1b-a939-47e2-84aa-2113366b0855.png">
4. After clicking on that action, the output is:
<img width="1206" alt="image" src="https://user-images.githubusercontent.com/62502207/203463069-80bfe2dc-f3d0-4fc8-b0ee-4889738ae0af.png">

Now, here are some thoughts. Note that I'm a Rust beginner (I encountered this while going though the Rust book hehe) so please take it with a grain of salt <3.
1. I don't think that code lens inside code that I didn't write myself should be enabled at all, since those feel like "implementor/dev-only" tools, and I don't immediately see why the clients of the APIs would need them. So I wouldn't be upset if this feature is removed (it's not working anyway...).
2. From the picture included in point 4 above, it seems like `cargo test` is running inside my current workspace, and not where the code lens is being displayed.

**Comment by MariaSolOs** [comment]
> I don't think that code lens inside code that I didn't write myself should be enabled at all, since those feel like "implementor/dev-only" tools, and I don't immediately see why the clients of the APIs would need them. So I wouldn't be upset if this feature is removed (it's not working anyway...).

This might be VS Code's behaviour, since the `Debug` code lens also happens inside `node_modules`:
<img width="1088" alt="image" src="https://user-images.githubusercontent.com/62502207/203464557-d0bbf870-9c51-496a-bc4c-f4f639bb0ea3.png">

**Comment by jedbrown** [comment]
I have a slightly different manifestation of what seems like the same bug, at top level `lib.rs`:
```rust
use std::marker::PhantomData;

pub struct X<'a> {
    field: PhantomData<&'a ()>,
}

impl<'a> X<'a> {
    /// ```
    /// analyzer_bug::X::new();
    /// ```
    pub fn new() -> Self {
        Self { field: PhantomData }
    }
}
```
The `Run Doctest` lens attempts
```
cargo test --doc --package analyzer-bug -- X::new --nocapture
```
when it should have used `"X<'a>::new"`.

Tested with `rust-analyzer 1.68.0-nightly (8a97b48 2022-12-22)`.
